python/ray/util/client/logsclient.py

```diff
@@ -18,47 +21,68 @@
+        self.client_worker = client_worker
         self._metadata = metadata
         self.request_queue = queue.Queue()
         self.log_thread = self._start_logthread()
         self.log_thread.start()
+        self.last_req = None
 
     def _start_logthread(self) -> threading.Thread:
         return threading.Thread(target=self._log_main, args=(), daemon=True)
 
     def _log_main(self) -> None:
-        stub = ray_client_pb2_grpc.RayletLogStreamerStub(self.channel)
-        log_stream = stub.Logstream(
-            iter(self.request_queue.get, None), metadata=self._metadata)
-        try:
-            for record in log_stream:
-                if record.level < 0:
-                    self.stdstream(level=record.level, msg=record.msg)
-                self.log(level=record.level, msg=record.msg)
-        except grpc.RpcError as e:
-            if e.code() == grpc.StatusCode.CANCELLED:
-                # Graceful shutdown. We've cancelled our own connection.
-                logger.info("Cancelling logs channel")
-            elif e.code() in (grpc.StatusCode.UNAVAILABLE,
-                              grpc.StatusCode.RESOURCE_EXHAUSTED):
-                # TODO(barakmich): The server may have
-                # dropped. In theory, we can retry, as per
-                # https://grpc.github.io/grpc/core/md_doc_statuscodes.html but
-                # in practice we may need to think about the correct semantics
-                # here.
-                logger.info("Server disconnected from logs channel")
-            else:
-                # Some other, unhandled, gRPC error
-                logger.exception(
-                    f"Got Error from logger channel -- shutting down: {e}")
+        reconnecting = False
+        while not self.client_worker._in_shutdown:
+            if reconnecting:
+                if log_once("lost_reconnect_logs"):
+                    logger.warning(
+                        "Log channel is reconnecting. Logs produced while "
+                        "the connection was down can be found on the head "
+                        "node of the cluster in "
+                        "`ray_client_server_[port].out`")
+                # Refresh queue and retry last request
+                self.request_queue = queue.Queue()
+                if self.last_req:
+                    self.request_queue.put(self.last_req)
+            stub = ray_client_pb2_grpc.RayletLogStreamerStub(
+                self.client_worker.channel)
+            try:
+                log_stream = stub.Logstream(
+                    iter(self.request_queue.get, None),
+                    metadata=self._metadata)
+            except ValueError:
+                # Trying to use the stub on a cancelled channel will raise
+                # ValueError. This should only happen when the data client
+                # is attempting to reset the connection -- sleep and try
+                # again.
+                time.sleep(.5)
```

Comment:
probably add a log message

Reply:
This is essentially the same as any other RPC error (which we currently don't warn on), just that the implementers decided to use ValueError instead of grpc.RpcError. IMO it's fine to leave out the warning

Reply:
Should we back off exponentially here?